Mirabuf Caching

a-crowell opened the pull request:
### Description
Utilizing browser storage for mirabuf caching: local storage for a map of keys and metadata, and OPFS for storing mirabuf data

### Objectives
- [X] Cache from fetchLocation or ArrayBuffer (if not already in)
- [X] Get assembly from ID
- [X] Remove specific entry from local storage and OPFS
- [X] Clear storage
- [X] SpawningModal separates mirabuf files in cache vs remote as well as uses the names of the files.

### Testing
Two new temp buttons on the MainHUD:
- Print Mira Maps -> prints the maps of robot/field fetch locations and ids in console
- Clear Mira -> clears both these maps stored in local storage and any mirabuf files stored in OPFS

Other than making sure robots and fields import correctly (no user difference from before), you can also check:
- Pulling from cache when it can and remote when it needs to (via console "Mira in cache" vs "Caching new mira")
- Saving correctly (via "Print Mira Maps" for the local storage maps and an OPFS explorer for the mirabuf files)
> <img src="https://github.com/Autodesk/synthesis/assets/53123817/04c444a3-b7bd-4af6-a2d7-bfbb258283e6" width="350">                    <img src="https://github.com/Autodesk/synthesis/assets/53123817/d8855ef8-ccde-40ff-985a-591d6207216f" width="400">

> [!Warning]
> Clear browser cache completely before testing. Storage structure is different now, so past data is breaking.

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1667)

a-crowell committed:
Readwrite with OPFS after localstorage check

a-crowell committed:
Quick save: ts update, clear opfs, robot/field dirs

a-crowell committed:
Catch for no Mira JSON in localStorage

a-crowell committed:
Working cache

a-crowell committed:
Function to get json object of map

a-crowell committed:
Misnomer correction and print maps prints both robots and fields

a-crowell committed:
Refactoring

a-crowell committed:
Merge remote-tracking branch 'origin/dev' into crowela/1667/mira-caching

a-crowell committed:
Merge conflicts part 2

BrandonPacewic commented on fission/src/Synthesis.tsx:
```diff
@@ -95,6 +95,11 @@ function Synthesis() {
                 .catch(_ => LoadMirabufRemote(DEFAULT_MIRA_PATH))
                 .catch(console.error)
 
+			const miraAssembly = await LoadMirabufRemote(mira_path, MiraType.ROBOT)
+				.catch(
+					_ => LoadMirabufRemote(DEFAULT_MIRA_PATH, MiraType.ROBOT)
+				).catch(console.error);
```
There seems to be a left over merge conflict here. Duplicate definition of `miraAssembly`.

BrandonPacewic commented on fission/src/ui/modals/spawning/SpawningModals.tsx:
```diff
@@ -7,7 +7,8 @@ import { useModalControlContext } from "@/ui/ModalContext"
 import { CreateMirabufFromUrl } from "@/mirabuf/MirabufSceneObject"
 import World from "@/systems/World"
-import { useTooltipControlContext } from "@/ui/TooltipContext"
+import { useTooltipControlContext } from "@/TooltipContext"
```
```suggestion
import { useTooltipControlContext } from "@/ui/TooltipContext"
```

Sneaky merge conflict.

BrandonPacewic reviewed (commented):
Couple of fixes after the merge `dev` commit that I had to make to run the branch.

HunterBarclay commented:
@BrandonPacewic Could you commit those fixes?

BrandonPacewic committed:
Update fission/src/ui/modals/spawning/SpawningModals.tsx

BrandonPacewic committed:
Merge fix

BrandonPacewic commented:
> @BrandonPacewic Could you commit those fixes?

@HunterBarclay, committed.

a-crowell committed:
Merge branch 'dev' into crowela/1667/mira-caching

a-crowell committed:
Merge fix

a-crowell committed:
Formatting

PepperLola reviewed (changes requested):
This looks good and seems like it works well. The only nitpicks I have are to add some documentation to the UnzipMira function (what do the magic numbers mean?) and then to use `undefined` instead of `null` per the [official TypeScript style guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined).

a-crowell committed:
Undefined instead of null

a-crowell reviewed (commented):
@PepperLola Done on the undefined. I didn't write the UnzipMira though. @HunterBarclay do you know how it works?

a-crowell committed:
Merge branch 'dev' into crowela/1667/mira-caching

a-crowell committed:
Merge fix

a-crowell committed:
UnzipMira doc

BrandonPacewic reviewed (commented):
Is this the correct behaviour, I have a robot and field in the scene. Robot data looks correct to me. For some reason the field data includes the blob of the local host? Is this correct or am I misstesting?

![image](https://github.com/Autodesk/synthesis/assets/92102436/e57883db-0e9c-4ab5-b941-0f35012ad061)

a-crowell commented:
@BrandonPacewic I have never seen that before. 1. What were the steps you took before then and 2. When was the last time you cleared browser history?

BrandonPacewic commented:
> @BrandonPacewic I have never seen that before. 1. What were the steps you took before then and 2. When was the last time you cleared browser history?

@a-crowell 

Steps:
- Loaded synthesis normally
- Printed the mira maps (normal behaviour)
- Spawned the field
- Printed the maps again (detailed behaviour)

I don't think I have ever cleared my browser history on this computer, what affects are we expecting / accounting on from that?

a-crowell commented:
> Steps:
> 
> * Loaded synthesis normally
> * Printed the mira maps (normal behaviour)
> * Spawned the field
> * Printed the maps again (detailed behaviour)
> 
> I don't think I have ever cleared my browser history on this computer, what affects are we expecting / accounting on from that?

@BrandonPacewic 

If the cache hadn't been cleared, the blob could have been leftover from an old review. But the maps printed empty just a few seconds earlier?

a-crowell commented:
Out of ideas for the localhost blob in local storage and unable to recreate it on any laptops (mine, Brandon's, Dhruv's) even checking for first time running this code, so leaving be for now.

HunterBarclay commented:
@BrandonPacewic I used a `URL.createObjectURL()` for the `ImportLocalMirabufModal` to convert the blob I get from the file input to a downloadable URL. We should probably customize the key for those, maybe a hash of the blob?

HunterBarclay commented:
@a-crowell ^^^

BrandonPacewic commented on fission/src/mirabuf/MirabufLoader.ts:
```diff
@@ -1,26 +1,125 @@
 import Pako from "pako"
-// import * as fs from "fs"
+
+const robots = "Robots"
+const fields = "Fields"
```
I would change these variable names to indicate what they are doing. Like `robotsDir` or something like that. If there not doing anything special and it would just be `robotText` then we could probably just leave them as string literals.

BrandonPacewic commented on fission/src/mirabuf/MirabufLoader.ts:
```diff
@@ -1,26 +1,125 @@
+    window.localStorage.removeItem(fields)
+}
+
+export function GetMap(type: MiraType): any {
```
Formatter getting upset at `any` here. Is this something we can change?

PepperLola replied:
I would create a type for the map. Something like `type MiraCache = { [path: string]: string }`, that way `GetMap` could return `MiraCache | undefined`. I also think `GetMap` could have a more descriptive name like `GetCache` because it's not immediately obvious what it does without reading the code.

BrandonPacewic commented on fission/src/mirabuf/MirabufLoader.ts:
```diff
@@ -1,26 +1,125 @@
+        console.log(assembly)
+        return assembly
+    } catch (e) {
+        console.error("Failed to load and cache mira")
```
I would definitely suggest also console logging the error and returning `undefined`

BrandonPacewic commented on fission/src/mirabuf/MirabufLoader.ts:
```diff
@@ -1,26 +1,125 @@
+            return assembly
+        }
+    } catch (e) {
+        console.error("Failed to find file from OPFS")
```
Same as previous except this time just log the error.

BrandonPacewic commented on fission/src/test/MirabufParser.test.ts:
```diff
@@ -2,12 +2,11 @@ import { describe, test, expect } from "vitest"
 import MirabufParser, { RigidNodeReadOnly } from "../mirabuf/MirabufParser"
-import { LoadMirabufRemote } from "@/mirabuf/MirabufLoader"
-// import { LoadMirabufLocal } from "../mirabuf/MirabufLoader"
+import { LoadMirabufRemote, MiraType } from "../mirabuf/MirabufLoader"
```
Reason for this not being imported like it was previously with `@`? Not sure why these are all over the place right now but this is not the only case so perhaps there is something that I am missing?

PepperLola replied:
I prefer using the @ path aliases because it means if we were to refactor and move mirabuf stuff somewhere else (or something like that) we would only have to change where the path alias points to rather than changing all the imports.

BrandonPacewic commented on fission/src/systems/simulation/behavior/GenericArmBehavior.ts:
```diff
@@ -11,6 +11,8 @@ class GenericArmBehavior extends Behavior {
     
     private _rotationalSpeed = 6;
 
+    private _rotationalSpeed = 30
+
```
Not sure where this change came from, perhaps another merge error?

BrandonPacewic reviewed (changes requested):
I'm impressed you were able to figure this out, very weird thing to get working correctly.

a-crowell committed:
Hashing and caching funky fetchLocations

a-crowell committed:
Code clarity and hashing bug fix

a-crowell committed:
Merge remote-tracking branch 'origin/dev' into crowela/1667/mira-caching

PepperLola commented:
What's the purpose of hashing the mira data?

a-crowell commented:
@PepperLola The caching usually compares the url/fetchLocation to determine whether or not the file is in the cache already. The ImportLocalMiraModal uses createObjectURL to pass in a fetchLocation though, giving keys of "blob: ... localhost ...", so instead, we now use the hash as a key.

PepperLola commented:
> @PepperLola The caching usually compares the url/fetchLocation to determine whether or not the file is in the cache already. The ImportLocalMiraModal uses createObjectURL to pass in a fetchLocation though, giving keys of "blob: ... localhost ...", so instead, we now use the hash as a key.

If you have to load, unzip, and decode the mira to hash it before checking if it's already in the cache, wouldn't that defeat the point of caching it in the first place? Or is there something I'm missing?

a-crowell commented:
> If you have to load, unzip, and decode the mira to hash it before checking if it's already in the cache, wouldn't that defeat the point of caching it in the first place? Or is there something I'm missing?

You're right what is the point of this.

In case Hunter meant hash of the url, I tried that too, but the whole thing is that it's different every time so it doesn't help either.

The two options I see are to either use the file name or just accept a new file from that modal and not cache it. @HunterBarclay You were against cutting down the fetchLocation to the name for other files--do your reasons still apply here?

Is there another way to check for same files I'm not thinking of?

BrandonPacewic reviewed (commented):
Thanks for making all my suggestions. It sounds like there is still some internal discussion that needs to be had on this feature so I'll wait to properly review it till after everything has been finalized.

a-crowell committed:
Refactor for readability and efficiency

Co-authored-by: Hunter Barclay <HunterBarclay@users.noreply.github.com>

a-crowell committed:
Bug fixes and more testing logs

a-crowell committed:
Name metadata, spawning modal update, and bug fixes

a-crowell committed:
Removed extraneous prints

a-crowell committed:
CacheAndGet

PepperLola reviewed (changes requested):
Instead of caching the file and then getting it and decoding there in the `CacheAndGetLocal`/`Remote` functions, why not unzip and decode there or create a new function to do it that is called everywhere else instead? If the end result of caching and then getting the cache is just getting the mira assembly, and if they're both using the same file buffer, this would skip however much time it takes to (unnecessarily?) read the file back from OPFS. Obviously let me know if I'm missing something, but to me it looks like a redundant step that could be skipped to save time.

a-crowell committed:
Improved CacheAndGet

a-crowell commented:
@PepperLola Thanks for pointing that out! I removed the remote version since it was never used anyway and adjusted `CacheAndGetLocal` to avoid grabbing from OPFS while taking advantage of having the assembly already to cache the name.

PepperLola reviewed (approved):
Looks good

BrandonPacewic reviewed (commented):
I'm not 100% up to date on this code and I just want to make sure that @HunterBarclay 's comment was address / is no longer relevant.

> I used a `URL.createObjectURL()` for the `ImportLocalMirabufModal` to convert the blob I get from the file input to a downloadable URL. We should probably customize the key for those, maybe a hash of the blob?

a-crowell commented:
@BrandonPacewic Yep, all covered! That modal now uses a hash of the file as key instead of url, so you won't find another localhost in the map.

BrandonPacewic committed:
Push formatting

BrandonPacewic reviewed (approved):
I tried to push formatting fixes but this branch is still not passing validation for some reason. I mentioned this in another PR review.

From what I can tell looks good, will be easier to test everything when we have end to end functionality.

HunterBarclay committed:
Little cleanups

HunterBarclay commented:
Fixed the unit tests and added a few clean ups. Didn't format because it still doesn't work on my desktop.

HunterBarclay committed:
That took me way too long

HunterBarclay committed:
Adding a version printout of the version

HunterBarclay committed:
Manually changed line endings to CRLF on Prettier

HunterBarclay committed:
Changed it back

HunterBarclay commented:
Merging for now. This Prettier issue is going to be a continuous problem
